lib/improver/weights.py

```diff
@@ -30,123 +30,259 @@
+            raise ValueError(msg)
+
+        if len(weights) != len(missing_weights):
+            msg = ('Arrays weights and missing_weights not the same size'
+                   ' weights is len {0:}'.format(len(weights)) +
+                   ' missing_weights is len {0:}'.format(len(missing_weights)))
+            raise ValueError(msg)
+
+        missing_sum = missing_weights.sum()
+        if missing_sum > 0.0:
+            combined_weights = weights*missing_weights
+            if method == 'evenly':
+                missing_avg_weight = (1.0 - combined_weights.sum())/missing_sum
+                redistributed_weights = combined_weights + missing_avg_weight
+            elif method == 'proportional':
+                redistributed_weights = (
+                    WeightsUtilities.normalise_weights(combined_weights))
+            else:
+                msg = ('Unknown weights redistribution method'
+                       ': {}'.format(method))
+                raise ValueError(msg)
+            # Set missing values to -1.0
+            redistributed_weights[np.where(missing_weights == 0.0)] = -1.0
         else:
-            slope = (ynval - y0val)/(num_of_weights - 1.0)
+            redistributed_weights = weights
+        return redistributed_weights
+
+    @staticmethod
+    def nonlinear_weights(num_of_weights, cval):
+        """Create nonlinear weights.
+
+            Args:
+                num_of_weights : Positive Integer
+                                 Number of weights to create.
+
+                cval : Float
+                       greater than 0.0 but less than or equal to 1,0,
+                       to be used for the nonlinear weights function.
+                       1.0 = equal weights for all.
+
+                       Weights will be calculated as
+                         cval**(tval-1)/Sum(of all weights)
+                       tval is the value of 1 to num_of_weights,
+
+            Returns:
+                weights : array of weights, sum of all weights = 1.0
+
+        """
+        if not isinstance(num_of_weights, int) or num_of_weights <= 0:
+            msg = ('Number of weights must be integer > 0, '
+                   'num = {0:s}'.format(str(num_of_weights)))
+            raise ValueError(msg)
+        if cval <= 0.0 or cval > 1.0:
+            msg = ('cval must be greater than 0.0 and less '
+                   'than or equal to 1.0 '
+                   'cval = {0:s}'.format(str(cval)))
+            raise ValueError(msg)
+
+        weights_list = []
+        for tval_minus1 in range(0, num_of_weights):
+            weights_list.append(cval**(tval_minus1))
+
+        weights = WeightsUtilities.normalise_weights(np.array(weights_list))
 
-    weights_list = []
-    for tval in range(0, num_of_weights):
-        weights_list.append(slope*tval + y0val)
+        return weights
+
+    @staticmethod
+    def linear_weights(num_of_weights, y0val=1.0, slope=0.0,
```

Comment:
The default value of 0.0 will give equal weighting to all forecasts. I think this is an intentional feature so I will leave it.